Assets/Scripts/Player.cs

```diff
@@ -3,25 +3,92 @@
 using System.Collections.Generic;
 using UnityEngine;
 
+// aight so what we want? 
+    // if MC, move according to controls
+    // if !MC: 
+        // if outside MC following position, get there quicker than movementspeed
+        // if very near position, normal follow
+
 public class Player : MonoBehaviour
 {
-    public float movementSpeed = 1f;
-    public float rotationSpeed = 1f;
-    private CharacterController characterController;
-    
+    [SerializeField]
+    private bool IsMainPlayer = false;
+    [SerializeField]
+    private float MovementSpeed = 1f;
+    [SerializeField]
+    private float RotationSpeed = 1f;
+    [SerializeField]
+    private float FollowDistance = 5f;
+    [SerializeField]
+    private float CatchUpSpeed = 2f; // ? TODO
+
+    private Vector3 TargetPosition;
+    private Quaternion TargetRotation;
+    private Vector3 Direction;
+
+    private bool IsMovementDisabled;
+
+    private CharacterController CharacterController;
+    private Player MainPlayer;
+    private Timer Timer;
+
     // Start is called before the first frame update
     void Start()
     {
-        characterController = GetComponent<CharacterController>();
+        CharacterController = GetComponent<CharacterController>();
+        var foundPlayers = GameObject.FindObjectsOfType<Player>();
+        MainPlayer = Array.Find(foundPlayers, player => player.IsMainPlayer);
+        Timer = GetComponent<Timer>();
     }
 
     // Update is called once per frame
     void Update()
     {
-        UpdateMovement();
+        if (IsMovementDisabled)
+            return;
+
+        if (IsMainPlayer)
+            UpdateMainCharacterMovement();
+        else
+            UpdateSecondaryCharacterMovement();
+    }
+
+    private void UpdateSecondaryCharacterMovement()
+    {
+        if (!Timer.IsInProgress())
+        {
+            Timer.StartTimer(.01f);
+
+            if (MainPlayer == null)
+            {
+                var foundPlayers = GameObject.FindObjectsOfType<Player>();
```

Comment:
Ideally we should only use `find` once in an initialization method. Searching again for stuff every update is a waste of resources unless there's a change of targets (which is unlikely in this case since it's the player characters)

Reply:
I put this here for the possibility of Change of Targets actually. 

If the MC is null for whatever reason, is will fetch another one. This doesn't happen every frame

Reply:
Hmm but if we wanted to ever change targets, that's something that should be done in a single method call, not a check every update. I would advise to move that in the Start still, and if we ever have a change of target we'll call a method that will redo the check for players/main player.

Reply:
> This doesn't happen every frame

It's called every time the characters are moving though, right? That's still a significant amount of calls compared to how many times there would be a change.

Reply:
Well, it's a null check that starts the flow. Computationally, it's insignificant I believe.

Reply:
aah I missed that check my bad. It's indeed a small cost in total, but it still feels inelegant; stuff that can be called at known specific moments should be done so instead of being checked periodically. Even if it's small, this is still a periodic added check that doesn't need to be periodic in the first place. Feels like bad practice in the long run. Well I won't insist more than that though.

Reply:
No no, good point! Ideally we can call a SwitchFocus function somewhere and have it done on a per call basis. But, seeing as we have a few days left... mind if we leave for now? 😆

Reply:
haha that's why I decided to stop insisting 😂 Yes we can leave it be 😛